### PR TITLE
Add username validation adaptive config and modify validations

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
## Purpose
- Currently the behavior of the IDF handler is to validate username if the tenant contains associated organization (B2B). However this has altered the previous default behavior of not validating the user in the normal flow.
- Also prompting to resolve the user by setting the "ValidateUsername" doesn't try to resolve the user from secondary user stores.

## Goals
1. This fix will move the server level "ValidateUsername" authenticator config into the adaptive script parameter, where the server config will get overridden by the adaptive property. The user resolution behavior would be as follows.
    ```
    if skipIdentifierPreProcess == "true":
        skipAll;
    
    else:
        if emailUserNameValidationEnabled == true:
            validateUsername();
    
        if multiAttributeLoginEnabled == true:
            resolveUser();
    
        if ValidateUsername(adaptive) == true:
    
            if B2B tenant:
                resolveB2BUser();
    
            else:
                resolveUser();
    
        if ValidateUsername(adaptive) == false:
            // Skip user resolution
    
        if ValidateUsername(adaptive) == null && ValidateUsername(server) == true:
            resolveUser();
    ```

2. Resolve user from secondary userstores when the user validation is enabled and user not found in the PRIMARY userstore.

## Related Issues
- https://github.com/wso2/product-is/issues/16037

## Related Pull Requests
- https://github.com/wso2-extensions/identity-local-auth-basicauth/pull/145

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java [hunk 1]
```diff
@@ -87,6 +90,7 @@ public class IdentifierHandler extends AbstractApplicationAuthenticator
     private static final String CONTINUE = "continue";
     private static final String RESET = "reset";
     private static final String RE_CAPTCHA_USER_DOMAIN = "user-domain-recaptcha";
+    private static final String VALIDATE_USERNAME_ADAPTIVE = "ValidateUsername";
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java:93]:
```suggestion
    private static final String VALIDATE_USERNAME_ADAPTIVE_SCRIPT_PARAM = "ValidateUsername";
```

File: components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java [hunk 2]
```diff
@@ -406,110 +412,85 @@ protected void processAuthenticationResponse(HttpServletRequest request,
...
-                        userStoreManager = (AbstractUserStoreManager) userRealm.getUserStoreManager();
-
-                        // If the user id is already resolved from the multi attribute login, we can assume the user
-                        // exists. If not, we will try to resolve the user id, which will indicate if the user exists
-                        // or not.
-                        if (userId == null) {
-                            userId = userStoreManager.getUserIDFromUserName(tenantAwareUsername);
+        if (validateUsernameAdaptiveParam != null) {
+            // If the "ValidateUsername" adaptive parameter is set, it should be honoured regardless of the
+            // "ValidateUsername" authenticator config.
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java:423]: Let's make this a multiline comment

File: components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java [hunk 3]
```diff
@@ -406,110 +412,85 @@ protected void processAuthenticationResponse(HttpServletRequest request,
...
-
-                    if (userRealm != null) {
-                        userStoreManager = (AbstractUserStoreManager) userRealm.getUserStoreManager();
-
-                        // If the user id is already resolved from the multi attribute login, we can assume the user
-                        // exists. If not, we will try to resolve the user id, which will indicate if the user exists
-                        // or not.
-                        if (userId == null) {
-                            userId = userStoreManager.getUserIDFromUserName(tenantAwareUsername);
+        if (validateUsernameAdaptiveParam != null) {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java:421]: Let's use stringUtils to check not null

Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java:421]: Add a comment mentioning the requirement of checking the not null here. 
i.e if it is null need to give the priority to the server config.

File: components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java [hunk 4]
```diff
@@ -406,110 +412,85 @@ protected void processAuthenticationResponse(HttpServletRequest request,
...
+            // If the "ValidateUsername" adaptive parameter is set, it should be honoured regardless of the
+            // "ValidateUsername" authenticator config.
+
+            if (Boolean.parseBoolean(validateUsernameAdaptiveParam)) {
+                boolean userNameValidated = false;
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java:426]: userNameValidated > isUserNameValidationRequired

File: components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java (hunk 1 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java]: Fixed

File: components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java (hunk 3 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java]: Fixed

Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java]: Fixed

File: components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java (hunk 4 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/IdentifierHandler.java]: Fixed

**Review by DMHP — APPROVED** [review]

**Pull request merged**